index.js

```diff
@@ -138,82 +89,170 @@ ModeDevice.prototype.setUpApiHttp = function() {
+  this._wsCleanUp();
+  this._wsScheduleReconnect();
+};
+
+ModeDevice.prototype._wsHandleOpen = function() {
+  debuglog('Websocket connection is open');
+
+  this._wsReconnectAttempts = 0;
+  this._wsReconnectDelay = 0;
+
+  debuglog('Scheduling periodic websocket pings');
+  this._wsPingTimer = setInterval(this._wsPing.bind(this), ModeDevice.WS_PING_INTERVAL * 1000);
+};
+
+ModeDevice.prototype._wsHandleMessage = function(message) {
+  debuglog('Received websocket message: "%s"', message);
+  let data;
+  try {
+    data = JSON.parse(message);
+  } catch (e) {
+    debuglog('Message is invalid JSON');
+    return;
+  }
+
+  try {
+    this.commandCallback(data);
+  } catch (e) {
+    debuglog('Error in command callback: %s', e);
+  }
+};
+
+ModeDevice.prototype._wsPing = function() {
+  if (this._wsPendingPingMsg != null) {
+    // Did not receive pong for the previous ping.
+    debuglog('Did not receive websocket pong for ping (%s)', this._wsPendingPingMsg);
+    this._wsDisconnect(); // this will trigger auto reconnection.
+    return;
+  }
+
+  if (this._websocket) {
+    const msg = 'ts=' + Date.now();
+
+    try {
+      this._websocket.ping(msg);
+    } catch (e) {
+      debuglog('Failed to ping websocket: %s', e);
+      this._wsDisconnect(); // this will trigger auto reconnection.
+      return;
+    }
+
+    this._wsPendingPingMsg = msg;
+    debuglog('Sent websocket ping: %s', msg);
   }
-  this.websocket = null;
 };
 
-ModeDevice.prototype.reconnect = function() {
-  debuglog('Reconnecting websocket');
-  if (this.websocket != null) {
-    debuglog('there is an websocket');
-    this.close();
-    return;  // reconnecting will be triggered by close event handler.
+ModeDevice.prototype._wsHandlePong = function(msg) {
+  debuglog('Received websocket pong: %s', msg);
+  if (msg == this._wsPendingPingMsg) {
+    this._wsPendingPingMsg = null;
   }
+};
 
-  var proto = this.apiUseTls ? 'wss' : 'ws';
-  var target = proto + '://' + this.apiHost + ':' + this.apiPort + '/devices/' + this.deviceId + '/command';
+ModeDevice.prototype._wsConnect = function() {
+  const proto = this.apiUseTls ? 'wss' : 'ws';
+  const target = proto + '://' + this.apiHost + ':' + this.apiPort + '/devices/' + this.deviceId + '/command';
 
-  debuglog("Connecting to " + target);
-  this.websocket = new ws(target, {
+  debuglog('Making websocket connection to %s', target);
+  this._websocket = new ws(target, {
     agent: this.apiHttpAgent,
     headers: {
       "Authorization": 'ModeCloud ' + this.token
     }
   });
-  this.websocket.on('error', this.errorCallback.bind(this));
-  this.websocket.on('close', this.closeCallback.bind(this));
-  this.websocket.on('open', this.openCallback.bind(this));
-  this.websocket.on('message', commandHandler.bind(this));
-  this.websocket.on('pong', this.pongCallback.bind(this));
+
+  this._websocket.on('error', this._wsHandleError.bind(this));
```

Comment:
The reconnection logic is built into the error and close handler so they should not be overridden, or the whole thing may break. In a future version, we can probably turn this into an emitter and emit 'connected' and 'disconnected' events to notify the client of a change in the websocket state. But those should be informational only, and not meant to client to override the behavior of the library.